File: R/AmpPhaseDecomp.R
AmpPhaseDecomp <- function(xfd, yfd, wfd)
{
#  Computes the amplitude-phase decomposition for a registration.

#  Arguments:
#  XFD  ...  FD object for unregistered functions
#  YFD  ...  FD object for registered functions
#  Wfd  ...  FD object for W functions determining warping functions

#  Returns:
#  MS.amp ... mean square for amplitude variation 
#  MS.pha ... mean square for amplitude variation 
#  RSQR   ... squared correlation measure of prop. phase variation 
#  C      ... constant C

#  Last modified 24 November 2008

xbasis  <- xfd$basis
nxbasis <- xbasis$nbasis
nfine   <- 10*nxbasis
xrng    <- xbasis$rangeval
tfine   <- seq(xrng[1],xrng[2],len=nfine)
delta   <- tfine[2]-tfine[1]

wfine   <- eval.fd(tfine, wfd)
xfine   <- eval.fd(tfine, xfd)
yfine   <- eval.fd(tfine, yfd)
efine   <- exp(wfine)
mufine  <- apply(xfine, 1, mean)
etafine <- apply(yfine, 1, mean)
N       <- dim(xfine)[2]
rfine   <- yfine - outer(etafine,rep(1,N))

intetasqr <- delta*(sum(etafine^2)-0.5*(etafine[1]^2 + etafine[nfine]^2))
intmusqr  <- delta*(sum(mufine^2) -0.5*(mufine[1]^2  + mufine[nfine]^2))

Cnum <- matrix(0,nfine,1)
for (i in 1:nfine) {
    Dhi     <- efine[i,]
    Syi     <- yfine[i,]^2
    Cnum[i] <- cov(Dhi, Syi)
}
intCnum <- delta*(sum(Cnum)-0.5*(Cnum[1]+Cnum[nfine]))
intysqr <- rep(0,N)
intrsqr <- rep(0,N)
for (i in 1:N) {
    intysqr[i] <- delta*(sum(yfine[,i]^2)-0.5*(yfine[1,i]^2 + yfine[nfine,i]^2))
    intrsqr[i] <- delta*(sum(rfine[,i]^2)-0.5*(rfine[1,i]^2 + rfine[nfine,i]^2))
}
C      <- 1 + intCnum/mean(intysqr)
MS.amp <- C*mean(intrsqr)
MS.pha <- C*intetasqr - intmusqr
RSQR   <- MS.pha/(MS.amp+MS.pha)

return(list("MS.amp" = MS.amp, "MS.pha" = MS.pha, "RSQR" = RSQR, "C" = C)) 

}

File: inst/Matlab/fdaM/AmpPhaseDecomp.m
function [MS_amp, MS_pha, RSQR, C] = AmpPhaseDecomp(xfd, yfd, hfd, rng)
%  Computes the amplitude-phase decomposition for a registration.

%  Arguments:
%  XFD  ...  FD object for unregistered functions
%  YFD  ...  FD object for registered functions
%  Hfd  ...  FD object for warping functions
%  RNG  ...  Sub-interval over which the decomposition is computed

%  Returns:
%  MS_amp ... mean square for amplitude variation 
%  MS_pha ... mean square for amplitude variation 
%  RSQR   ... squared correlation measure of prop. phase variation 
%  C      ... constant C

%  Last modified 21 April 2009

%  extract information from XFD

xbasis  = getbasis(xfd);
xrng    = getbasisrange(xbasis);

%  set default interval RNG

if nargin < 4, rng = xrng;  end

%  check RNG

if rng(1) < xrng(1) || rng(2) > xrng(2)
    error('RNG values outide XRNG interval.');
end
if rng(1) >= rng(2)
    error('RNG values not strictly increasing.');
end

%  set up a fine mesh of values for numerical integration

nxbasis = getnbasis(xbasis);
nfine   = max([201,10*nxbasis]);
tfine   = linspace(rng(1),rng(2),nfine)';
delta   = tfine(2) - tfine(1);

%  evaluate arguments at fine mesh values

xfine   = eval_fd(tfine, xfd);
yfine   = eval_fd(tfine, yfd);
Dhfine  = eval_fd(tfine, hfd, 1);

%  means of unregistered and registered functions

mufine  = mean(xfine,2);
etafine = mean(yfine,2);

%  integrated squared mean values

intetasqr = delta.*trapz(etafine.^2);
intmusqr  = delta.*trapz( mufine.^2);

%  covariance between Dh and y values

covDhSy = zeros(nfine,1);
for i=1:nfine
    Dhi        = Dhfine(i,:);
    Syi        = yfine(i,:).^2;
    Covmat     = cov(Dhi', Syi');    
    covDhSy(i) = Covmat(1,2);
end

%  integrated covariance value

intcovDhSy = delta.*trapz(covDhSy);

%  integrated squared registered and registered residual values

N = size(xfine,2);
intysqr = zeros(N,1);
intrsqr = zeros(N,1);
rfine   = yfine - etafine*ones(1,N);
for i=1:N
    intysqr(i) = delta.*trapz(yfine(:,i).^2);
    intrsqr(i) = delta.*trapz(rfine(:,i).^2);
end

%  compute results

C      = 1 + intcovDhSy/mean(intysqr);
MS_amp = C*mean(intrsqr);
MS_pha = C*intetasqr - intmusqr;
RSQR   = MS_pha/(MS_amp+MS_pha);
